### ISSUE TITLE
AAT, I'd like a webview test that asserts (NO OFFSHORE) Check version.txt to make sure the latest code is deployed

### ISSUE THREAD

**Issue by m1yag1** [openstax/cnx-automation#150]
# CASE NUM: C175150

## User story

```
AAT, I'd like a webview test that asserts (NO OFFSHORE) Check version.txt to make sure the latest code is deployed
```

## Preconditions

```

```

## Steps

```
Go to https://staging.cnx.org/version.txt and check the date, and make sure the versions/shas/tags for the repos listed match the latest release for each repo.

NOTE: Be sure the newest tags are displaying on each `/tags` page listed below. Most recent tag should be listed at the top.

- Date should be very recent (within the last few days).
- Check webview version against https://github.com/Connexions/webview/tags
- Check cnx-archive version against https://github.com/Connexions/cnx-archive/tags
- Check cnx-deploy version against https://github.com/Connexions/cnx-deploy/tags


```

## Expected Result

```
The version/sha numbers on version.txt should match those for the latest tag in each repo.
```

**Comment by kerwinso** [comment]
You know what would be cool… if we could automatically parse `history.txt` and report somewhere what changed from the most recent deployment vs. the next most recent deployment. If all the hashes/pins match between the two deployments, then that’s a fail.

**Comment by Dantemss** [comment]
https://github.com/openstax/cnx-automation/pull/179

**Comment by Dantemss** [comment]
@kerwinso looking at https://staging.cnx.org/history.txt, it seems some entries are duplicated, for example `2018-06-15 13:20:36 CDT` and `2018-06-15 13:58:56 CDT` contain the same versions for everything.

It might be better to give warnings in this case too instead of failing (and maybe print a diff with the release before that one instead).

**Comment by kerwinso** [comment]
@Dantemss A warning in that case is fine. Also, would it be possible to print a diff every time whether it passes or fails?

**Comment by Dantemss** [comment]
Yes, we can print the diff regardless of failure (seems like if we're printing warnings for almost everything, the only failure for this test is if version.txt and history.txt disagree, which seems unlikely).

**Comment by kerwinso** [comment]
@Dantemss Well, we wouldn't be comparing version.txt to history.txt -- those always have the same content (version.txt is a subset of the data in history.txt). What I'm interested in diffing is the data for different deployment dates _within_ history.txt.

**Comment by Dantemss** [comment]
@kerwin well they SHOULD have the same content :)

Here's the output I got so far:

```
tests/webview/ui/test_version.py::test_version PASSED

----- generated html file: /Users/dante/Desktop/cnx-automation/report.html -----

Release diff from the previous release to the current release:

--- 
+++ 
@@ -1,15 +1,13 @@
-
-
-{"date": "2018-06-14 09:58:01 CDT",
- "webview": "c21f313c844f5053a997cbdf26c7818ba75a56a9 (v0.31.0-2-gc21f313)",
+{"date": "2018-06-21 15:45:43 CDT",
+ "webview": "1a986e03fd1c66d2337792a98aed0683e37fd68b (v0.31.1-3-g1a986e0)",
  "cnx-archive": "e916e59b1ca336b1449f6535f3f3c73593a4652a",
  "cnx-publishing": "14a120e31e216320d4352b6bd2df8cbcf37df577",
  "oer.exports": "kawaii.kitsune v0.56.0",
- "press": "3.3.0",
- "cnx-deploy": "v2.18.2-35-gf5e69e8"
+ "press": "3.3.1",
+ "cnx-deploy": "v2.18.2-57-g522c64b"
 }
 -------------------------------
-# 2018-06-14 09:58:02 CDT
+# 2018-06-21 15:45:45 CDT
 # Archive full versions from qa00.cnx.org:
 appdirs==1.4.0
 certifi==2017.4.17

=============================== warnings summary ===============================
tests/webview/ui/test_version.py::test_version
  /Users/dante/Desktop/cnx-automation/tests/webview/ui/test_version.py:56: UserWarning: 
  
  The cnx-deploy tag for the latest release in https://qa.cnx.org/history.txt (v2.18.2) does not match the latest tag in https://github.com/Connexions/cnx-deploy.git (v2.19.8)
    latest_tag=latest_tag

-- Docs: http://doc.pytest.org/en/latest/warnings.html
============ 1 passed, 133 deselected, 1 warnings in 16.54 seconds =============
flake8 installed: flake8==3.5.0,mccabe==0.6.1,pycodestyle==2.3.1,pyflakes==1.6.0
flake8 runtests: PYTHONHASHSEED='1629131288'
flake8 runtests: commands[0] | flake8
___________________________________ summary ____________________________________
  py36: commands succeeded
  flake8: commands succeeded
  congratulations :)
```

**Comment by kerwinso** [comment]
@Dantemss This looks pretty good. A few questions and comments:

1) Is there a 2 column diff option? Curious what that would look like.
2) Warnings rather than fails are good. 
3) I don't understand why these lines:
```
-# 2018-06-14 09:58:02 CDT
+# 2018-06-21 15:45:45 CDT
 # Archive full versions from qa00.cnx.org:
 appdirs==1.4.0
 certifi==2017.4.17
```
I may be misinterpreting, but it looks like the diff is telling us that `appdirs` and `certifi` changed between the deployments on 6/14 and 6/21. But they didn't -- look at https://qa.cnx.org/history.txt to verify.

**Comment by mikefromit** [comment]
where there is a `-` that is the previous and `+` that is the new. The other lines with `appdirs` and `certifi` have no changes due to no symbols beside them.

**Comment by Dantemss** [comment]
I haven't seen a 2-column option, but I'll look again.
The diff tool provides nearby lines as context, that's why you see appdirs and certifi. As Mike said, there are no changes to them.

**Comment by m1yag1** [comment]
where there is a `-` that is the previous and `+` that is the new. The other lines with appdirs and certifi have no changes due to no symbols beside them.

**Comment by Dantemss** [comment]
There is an option to make an HTML diff. I think this would have a table with the diff so it's kind of a 2-column option. Of course, the output would be on a separate file and you'd have to open it on your browser.

**Comment by m1yag1** [comment]
Let's stay away from the HTML diff. The diff format from the console may take some getting used to but is a good thing to learn and become proficient with, especially when using `git diff`. https://stackoverflow.com/questions/2529441/how-to-read-the-output-from-git-diff.

**Comment by m1yag1** [comment]
Resolved by #179
